client/src/session_layer.rs

```diff
@@ -0,0 +1,679 @@
+        };
+
+        let crypto = match self.config.crypto.get(self.config.node_id).await {
+            Ok(crypto) => crypto,
+            Err(e) => return Box::pin(futures::future::ready(Err(e))),
+        };
+
+        // Compute challenge in different thread to avoid blocking runtime.
+        // Note: computing starts here, not after awaiting.
+        challenge::solve_blocking(request.challenge, request.difficulty, crypto)
+    }
+}
+
+impl Handler for SessionsLayer {
+    fn dispatcher(&self, from: SocketAddr) -> LocalBoxFuture<Option<Dispatcher>> {
+        let handler = self.clone();
+        async move {
+            let state = handler.state.read().await;
+
+            // We get either dispatcher for already existing Session from self,
+            // or temporary session, that is during creation.
+            state
+                .sessions
+                .get(&from)
+                .map(|session| session.dispatcher())
+                .or_else(|| {
+                    state
+                        .starting_sessions
+                        .clone()
+                        .map(|entity| entity.dispatcher(from))
+                        .flatten()
+                })
+        }
+        .boxed_local()
+    }
+
+    fn on_control(
+        &self,
+        _session_id: Vec<u8>,
+        control: proto::Control,
+        from: SocketAddr,
+    ) -> LocalBoxFuture<()> {
+        log::debug!("Received control packet from {}: {:?}", from, control);
+        Box::pin(futures::future::ready(()))
+    }
+
+    fn on_request(
+        &self,
+        session_id: Vec<u8>,
+        request: proto::Request,
+        from: SocketAddr,
+    ) -> LocalBoxFuture<()> {
+        log::debug!("Received request packet from {}: {:?}", from, request);
+
+        let (request_id, kind) = match request {
+            proto::Request {
+                request_id,
+                kind: Some(kind),
+            } => (request_id, kind),
+            _ => return Box::pin(futures::future::ready(())),
+        };
+
+        match kind {
+            proto::request::Kind::Ping(request) => {
+                Box::pin(self.on_ping(session_id, request_id, from, request))
+            }
+            proto::request::Kind::Session(request) => {
+                Box::pin(self.dispatch_session(session_id, request_id, from, request))
+            }
+            _ => Box::pin(futures::future::ready(())),
+        }
+    }
+
+    fn on_forward(&self, forward: proto::Forward, from: SocketAddr) -> LocalBoxFuture<()> {
+        let myself = self.clone();
+        let fut = async move {
+            log::trace!(
+                "[{}] received forward packet ({} B) via {}",
+                myself.config.node_id,
+                forward.payload.len(),
+                from
+            );
+
+            let node = if forward.slot == 0 {
+                // Direct message from other Node.
+                match myself.state.read().await.nodes_addr.get(&from).cloned() {
```

Comment:
should we release the read lock asap?
```suggestion
                match {myself.state.read().await.nodes_addr.get(&from).cloned()} {
```